### PR TITLE
feat: add script to simplify the package release process (#27)

### PR DESCRIPTION
**Pull request by onebytegone**
This is a follow up for #28

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +115,95 @@ grunt.registerTask('standards', [ 'markdownlint' ]);
...
+      for the project in its git hosting solution, e.g.
+      <https://github.com/silvermine/event-emitter.git> for the
+      [@silvermine/event-emitter](https://github.com/silvermine/event-emitter) project.
+      * This is necessary because conventional-changelog needs to know the URL to the git
+         hosting solution so that it can make links to "compare URLs" in the CHANGELOG
+   1. Add the following NPM scripts to the project's `package.json` file:
+
+      ```json
+      "release:prep-changelog": "./node_modules/@silvermine/standardization/scripts/release.sh prep-changelog",
+      "release:finalize": "./node_modules/@silvermine/standardization/scripts/release.sh finalize"
```
Comment (pbredenberg) [inline README.md:134]: Looks like these would need to be updated to point to the Node versions of these scripts, if this implementation were to be accepted.

**Comment by yokuze** [comment]
I left a comment about this PR here: https://github.com/silvermine/standardization/pull/28#issuecomment-1055891474 that we should discuss first before I give a review.

File: README.md (hunk 1 above)
Comment (onebytegone) [reply README.md]: 🤦‍♂️  Good catch

**Review by yokuze — CHANGES_REQUESTED** [review]
Great work here @onebytegone! Big improvements over #28.

**Comment by yokuze** [comment]
@onebytegone feel free to rebase off of `master`. It won't mess up my subsequent review.

File: scripts/release.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      await prepareChangelog({
+         currentVersion,
+         targetVersion,
+         issueUrlFormat: changelogConfig ? changelogConfig.issueUrlFormat : undefined,
+      });
+   } else if (argv[0] === 'finalize') {
+      console.info(`Bumping ${packageJSON.name} from ${currentVersion} to ${targetVersion}...`);
+      await bumpPackageVersion(targetVersion);
+   } else {
+      throw Error(`ERROR: Unexpected input, received: ${JSON.stringify(argv)}`);
```
[the remark this answers is not in the document]
Comment (onebytegone) [reply scripts/release.js]: I think this is from the old version. Switching to commander should have made this obsolete.

File: scripts/release.js [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+         }
+
+         await addEntriesToChangelog(changelogAdditions);
+
+         await commitChangelog(targetVersion);
+
+         console.info(
+            chalk.whiteBright('The changelog has been updated. Please do the following:\n\n')
+            + `   ${chalk.gray('1.')} git show ${chalk.gray('# and review the changes')}\n`
+            + `   ${chalk.gray('2.')} git push origin ${changelogBranch}\n`
```
[the remark this answers is not in the document]
Comment (onebytegone) [reply scripts/release.js:251]: Good point. I'd like to make these commands a simple copy/paste thing if possible. What do you think about logging the remote that the source branch has as its upstream? (see example below) I think in most cases the branch that versions are being cut from should be tracking the correct remote. If there is no upstream set, we could fallback to a `$REMOTE_NAME` placeholder.

```
$ git remote -v
onebytegone	git@github.com:onebytegone/generator-typescript.git (fetch)
onebytegone	git@github.com:onebytegone/generator-typescript.git (push)
silvermine	git@github.com:silvermine/generator-typescript.git (fetch)
silvermine	git@github.com:silvermine/generator-typescript.git (push)
$ git config branch.master.remote
silvermine
```

File: scripts/release.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yokuze) [reply scripts/release.js]: Oops, yep. I had a pending review on the old version and I missed deleting this comment. Thanks

File: scripts/release.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (yokuze) [reply scripts/release.js:251]: Sounds like a great idea.

File: scripts/release.js [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            '--allow-same-version',
+            targetVersion,
+            '-m',
+            `chore: version bump: v${targetVersion}`,
+         ]);
+
+         console.info(
+            chalk.whiteBright('The package version has been bumped. Please do the following:\n\n')
+            + `   ${chalk.gray('1.')} git show ${chalk.gray('# and review the changes')}\n`
+            + `   ${chalk.gray('2.')} git push origin ${currentBranch} --follow-tags --dry-run ${chalk.gray('# and review the changes')}\n`
```
[the remark this answers is not in the document]
Comment (onebytegone) [reply scripts/release.js:280]: At least for me, it's been a combo of:

* _The Way (tm), it ain't broke._ A number of articles document  using `--follow-tags`. For me it's a case of "well, this works. I'm just going to push the tags and move on to other things".
* `--follow-tags` makes it so you don't need to know the tag name. This makes it easier so static documentation can just be copied and run.

Unless I'm missing something, specifying the branch and tag should be fine in this case.

File: scripts/release.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (onebytegone) [reply scripts/release.js:251]: Gaaah, using the origin from the source branch doesn't work for the changelog branches. Those may need to be pushed to a fork which is not the upstream remote of `master`. 🤦‍♂️ 

I'll just stick with `$REMOTE_NAME`. There's not a simple way around this one.

[the remark this answers is not in the document]
Comment (yokuze) [reply scripts/release.js:251]: Oh - I was thinking you'd use something like `'master' || 'main'` as the branch name.

File: scripts/release.js [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+         releaseType = 'prerelease';
+      } else {
+         releaseType = `pre${releaseType}`; // e.g. premajor, preminor, prepatch
+      }
+   }
+
+   return semver.inc(currentVersion, releaseType, prereleaseIdentifier);
+}
+
+async function getMostRecentNonPrereleaseTag() {
```
[the remark this answers is not in the document]
Comment (onebytegone) [reply scripts/release.js]: [`git-semver-tags` runs a `git log` command](https://github.com/conventional-changelog/conventional-changelog/blob/21b73f974adc4490a318d9c892055655cb82c538/packages/git-semver-tags/index.js#L7) for the current commit. This means only tags that are for ancestors of the current commit will be listed and v2.0.0 wouldn't be returned.

An example:

```bash
$ git log --tags --no-walk --simplify-by-decoration --pretty="format:%ci %d" # All tags in the repo
2022-03-08 15:15:33 -0500  (tag: v2.0.1) # This is a merge of v1.2.1 into v2.0.0
2022-03-08 15:11:01 -0500  (tag: v1.2.1)
2022-03-08 15:10:15 -0500  (tag: v2.0.0)
2022-03-08 15:10:03 -0500  (tag: v1.2.0)
2022-03-08 15:09:50 -0500  (tag: v1.1.0)
2022-03-08 15:09:31 -0500  (tag: v1.0.0)
$ git checkout v1.2.1
$ git log --simplify-by-decoration --pretty="format:%ci %d" # Only list the tags that are on ancestors
2022-03-08 15:11:01 -0500  (HEAD, tag: v1.2.1)
2022-03-08 15:10:03 -0500  (tag: v1.2.0)
2022-03-08 15:09:50 -0500  (tag: v1.1.0)
2022-03-08 15:09:31 -0500  (tag: v1.0.0)
$ node
> const gitSemverTags = require('git-semver-tags');
undefined
> gitSemverTags({ skipUnstable: true }, console.log);
undefined
> null [ 'v1.2.1', 'v1.2.0', 'v1.1.0', 'v1.0.0' ]
```

However, there was a bug here. We need to sort the list of tags by semver. `git-semver-tags` returns the tags ordered by time. This meant when creating v2.0.1, the "last" version was v1.2.1 not v2.0.0. 

```bash
$ git checkout v2.0.1
$ git log --simplify-by-decoration --pretty="format:%ci %d" # Only list the tags that are on ancestors
2022-03-08 15:15:33 -0500  (HEAD, tag: v2.0.1)
2022-03-08 15:11:01 -0500  (tag: v1.2.1)
2022-03-08 15:10:15 -0500  (tag: v2.0.0)
2022-03-08 15:10:03 -0500  (tag: v1.2.0)
2022-03-08 15:09:50 -0500  (tag: v1.1.0)
2022-03-08 15:09:31 -0500  (tag: v1.0.0)
$ node
> const gitSemverTags = require('git-semver-tags');
undefined
> gitSemverTags({ skipUnstable: true }, console.log);
undefined
> null [ 'v2.0.1', 'v1.2.1', 'v2.0.0', 'v1.2.0', 'v1.1.0', 'v1.0.0' ]
> const semver = require('semver');
> gitSemverTags({ skipUnstable: true }, (err, tags) => { console.log(semver.rsort(tags)); });
undefined
> [ 'v2.0.1', 'v2.0.0', 'v1.2.1', 'v1.2.0', 'v1.1.0', 'v1.0.0' ]
```

Thanks for keeping on this. There was a bug afoot!

[the remark this answers is not in the document]
Comment (yokuze) [reply scripts/release.js]: Ah, thank you. That makes sense.

**Pull request merged**